edgelet/kube-client/src/client.rs

```diff
@@ -0,0 +1,441 @@
+        &mut self,
+        mut req: http::Request<Vec<u8>>,
+    ) -> impl Future<Item = http::Response<Body>, Error = Error> {
+        self.config
+            .host()
+            .join(self.config.api_path())
+            .and_then(|base_url| {
+                base_url.join(
+                    req.uri()
+                        .path_and_query()
+                        .map(|pq| pq.as_str())
+                        .unwrap_or(""),
+                )
+            })
+            .map_err(Error::from)
+            .and_then(|url| url.as_ref().parse().map_err(Error::from))
+            .and_then(|uri| self.config.token_source().get().map(|token| (uri, token)))
+            .and_then(|(uri, token)| {
+                // set the full URL on the request including API path
+                *req.uri_mut() = uri;
+
+                // add the authorization bearer token to the request if we have one
+                if let Some(token) = token {
+                    let token = format!("Bearer {}", token).parse()?;
+                    req.headers_mut().append(http::header::AUTHORIZATION, token);
+                }
+
+                Ok(req)
+            })
+            .map(|req| {
+                // NOTE: The req.map call below converts from Request<Vec<u8>> into a
+                // Request<Body>. The res.map call converts from S::ResBody to Body.
+                Either::A(
+                    self.client
+                        .call(req.map(From::from))
+                        .map_err(Into::into)
+                        .map(|res| res.map(From::from)),
+                )
+            })
+            .unwrap_or_else(|err| Either::B(future::err(err)))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use super::*;
+    use config::{Config, TokenSource};
+    use hyper::service::service_fn;
+    use hyper::{Body, Error as HyperError, Request, Response, StatusCode};
+    use k8s_openapi::v1_10::api::apps::v1 as apps;
+    use k8s_openapi::v1_10::apimachinery::pkg::apis::meta::v1 as api_meta;
+    use native_tls::TlsConnector;
+    use serde_json;
+    use tokio::runtime::Runtime;
+    use url::Url;
+
+    #[derive(Clone)]
+    struct TestTokenSource();
+
+    impl TokenSource for TestTokenSource {
+        type Error = super::Error;
+
+        fn get(&self) -> Result<Option<String>, Self::Error> {
+            Ok(None)
+        }
+    }
+
+    const STATUS_SUCCESS: &str = r###"{"kind":"Status", "status":"Success"}"###;
+    const DEPLOYMENT_JSON: &str = r##"{"kind":"Deployment"}"##;
+
+    #[test]
+    fn create_deployment_success() {
+        const NAMESPACE: &str = "custom-namespace";
+        let service1 = service_fn(
+            move |req: Request<Body>| -> Result<Response<Body>, HyperError> {
+                {
+                    let p = req.uri().path();
+                    assert!(p.contains(NAMESPACE));
+                    let q = req.uri().query().unwrap();
+                    assert!(q.is_empty());
+                }
```

Comment:
Switching to "Rust 2018" in `Cargo.toml` might help kick in the non-lexical lifetime sweetness which might obviate the need for these extra braces.

Reply:
Yes, I couldn't do "into_body()" because I was borrowing from the struct in these lines.  I added the braces so p and q would be dropped before I made the next call.

Updated crate to 2018 edition and was able to remove.

Reply:
:+1: